plugins/s3_auth/s3_auth.cc

```diff
@@ -1053,19 +1053,18 @@ int
 config_reloader(TSCont cont, TSEvent event, void *edata)
 {
   Dbg(dbg_ctl, "reloading configs");
-  S3Config *s3          = static_cast<S3Config *>(TSContDataGet(cont));
+  S3Config *s3 = static_cast<S3Config *>(TSContDataGet(cont));
+  std::unique_lock lock(s3->reload_mutex);
```

Comment:
It looks like this function is calling many `S3Config` functions that modifies the `s3` object, so acquiring the `reload_mutex` in the beginning seems better.

Reply:
This is a shared mutex.  It only protects the data changed by copy_changes_from().

Reply:
Oh! I thought this is regular mutex...it'll be not simple fix. I'll rollback this change and let races go for now.

Reply:
For the race, @moonchen is writing up the issue and approaches.